mongo_connector/dest_mapping.py

```diff
@@ -0,0 +1,186 @@
+                map_arr.append(x)
+            else:
+                map_arr.append(src_match.group(map_index))
+                map_index += 1
+        
+        return ''.join(map_arr)
+        
+    
+    def get(self,plain_src_ns,defaultval=""):
+        """Given a plain source namespace, return a mapped namespace if matched.
+        If no match and given defaultval, return it as a default value.
+        """
+        self.lock.acquire()
+        # search in plain mappings first
+        if plain_src_ns in self.plain.keys():
+            self.lock.release()
+            return self.plain[plain_src_ns]
+        else:
+            # search in wildcard mappings
+            # if matched, get a replaced mapped namespace
+            # and add to the plain mappings            
+            for k, v in self.wildcard.items():
+                m_res = self.match(plain_src_ns, k)
+                if m_res:
+                    res = self.replace(m_res, v)
+                    if res is not None:
+                        if res not in self.reverse_plain:
+                            self.plain[plain_src_ns] = res
+                            self.reverse_plain[res] = plain_src_ns
+                            
+                            # add matched plain db to plain_db dict, but skip $cmd case
+                            db, col = plain_src_ns.split(".", 1)
+                            if col != "$cmd":
+                                arr = self.plain_db.get(db, [])
+                                arr.append(res.split(".")[0])
+                                self.plain_db[db] = list(set(arr))
+                            self.lock.release()
+                            return res
+                        else:
+                            self.lock.release()
+                            raise errors.InvalidConfiguration("Destination namespaces set wildcard pattern conflicts with the plain pattern.")
+            
+        if defaultval:
+            self.lock.release()
+            return defaultval
+        else:
+            self.lock.release()
+            LOG.error("Failed to find matched mapping.")
+            return None
+    
+
+    def set(self, key, value):
+        """Set the DestMapping instance to corresponding dictionary"""
+        self.lock.acquire()
+        db = key.split(".")[0]
+        if "*" in db:
+            raise errors.InvalidConfiguration("Wildcard is not allowed in db name")
+            
+        if "*" in key:
+            if key.count("*") != value.count("*"):
+                raise errors.InvalidConfiguration("The number of wildcards in the mapping should be equal")
+            self.wildcard[key] = value
+        else:
+            if value not in self.reverse_plain:
+                self.plain[key] = value
+                self.reverse_plain[value] = key
+                arr = self.plain_db.get(db, [])
+                arr.append(value.split(".")[0])
+                self.plain_db[db] = list(set(arr))
```

Comment:
Perhaps self.plain_db should be a mapping of strings -> set objects. When adding something to self.plain_db[db], let's just call `self.plain_db[db].add(<something>)`